src/library/scala/collection/mutable/HashTable.scala

```diff
@@ -357,14 +360,14 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
 
   protected def elemEquals(key1: A, key2: A): Boolean = (key1 == key2)
 
-  // Note:
-  // we take the most significant bits of the hashcode, not the lower ones
-  // this is of crucial importance when populating the table in parallel
-  protected final def index(hcode: Int) = {
+  /**
+    * Note: we take the most significant bits of the hashcode, not the lower ones
+    * this is of crucial importance when populating the table in parallel
+    */
+  protected final def index(hcode: Int): Int = {
     val ones = table.length - 1
-    val improved = improve(hcode, seedvalue)
-    val shifted = (improved >> (32 - java.lang.Integer.bitCount(ones))) & ones
-    shifted
+    val exponent = Integer.numberOfLeadingZeros(ones)
+    (improve(hcode, seedvalue) >>> exponent) & ones
```

Comment:
```scala
(Integer.MAX_VALUE >>> 32 & 0) != (Integer.MAX_VALUE >>> 32)
```
i.e. if the length is `1`, `ones` will be `0`, `exponent` will be `32`, which would produce the same shift as `0` (i.e. `123 >>> 32` == `123`) ... other than that (i.e. if `length` > 1) this seems to be true indeed.
Should I simply assert that the `length >= 2` at the beginning?